src/smt/theory_str_noodler/decision_procedure.cpp

```diff
@@ -100,67 +170,53 @@ namespace smt::noodler {
-
-                for (; right_var_it != right_side_end; ++right_var_it) {
-                    std::shared_ptr<Mata::Nfa::Nfa> right_var_aut = element_to_process.aut_ass.at(*right_var_it);
-                    if (element_to_process.length_sensitive_vars.count(*right_var_it) > 0) {
-                        // current right_var is length-aware
+            assert(!right_side_vars.empty());
+            auto right_var_it = right_side_vars.begin();
+            auto right_side_end = right_side_vars.end();
+
+            std::shared_ptr<Mata::Nfa::Nfa> next_aut = element_to_process.aut_ass[*right_var_it];
+            std::vector<BasicTerm> next_division{ *right_var_it };
+            bool last_was_length = (element_to_process.length_sensitive_vars.count(*right_var_it) > 0);
+            ++right_var_it;
+
+            for (; right_var_it != right_side_end; ++right_var_it) {
+                std::shared_ptr<Mata::Nfa::Nfa> right_var_aut = element_to_process.aut_ass.at(*right_var_it);
+                if (element_to_process.length_sensitive_vars.count(*right_var_it) > 0) {
+                    // current right_var is length-aware
+                    right_side_automata.push_back(next_aut);
+                    right_side_division.push_back(next_division);
+                    next_aut = right_var_aut;
+                    next_division = std::vector<BasicTerm>{ *right_var_it };
+                    last_was_length = true;
+                } else {
+                    // current right_var is not length-aware
+                    if (last_was_length) {
                         right_side_automata.push_back(next_aut);
                         right_side_division.push_back(next_division);
                         next_aut = right_var_aut;
                         next_division = std::vector<BasicTerm>{ *right_var_it };
-                        last_was_length = true;
-                        is_there_length_on_right = true;
                     } else {
-                        // current right_var is not length-aware
-                        if (last_was_length) {
-                            right_side_automata.push_back(next_aut);
-                            right_side_division.push_back(next_division);
-                            next_aut = right_var_aut;
-                            next_division = std::vector<BasicTerm>{ *right_var_it };
-                        } else {
-                            next_aut = std::make_shared<Mata::Nfa::Nfa>(Mata::Nfa::concatenate(*next_aut, *right_var_aut));
-                            next_division.push_back(*right_var_it);
-                            // TODO probably reduce size here
-                        }
-                        last_was_length = false;
+                        next_aut = std::make_shared<Mata::Nfa::Nfa>(Mata::Nfa::concatenate(*next_aut, *right_var_aut));
+                        next_division.push_back(*right_var_it);
+                        // TODO probably reduce size here
```

Comment:
Could we try to resolve this forgotten TODO now?

Reply:
I think the redution is done during noodlification now, so it shouldn't be needed to do the reduction there. I am not completely sure, I am planning to go trough all TODOs during refactoring, maybe we can talk about the code on the next string meeting?